fedimint-core/src/encoding/collections.rs

```diff
@@ -0,0 +1,342 @@
+        let len = u64::consensus_decode_from_finite_reader(d, modules)?;
+        for _ in 0..len {
+            let k = K::consensus_decode_from_finite_reader(d, modules)?;
+            if res.last().is_some_and(|prev_key| k <= *prev_key) {
+                return Err(DecodeError::from_str("Non-canonical encoding"));
+            }
+            if !res.insert(k) {
+                return Err(DecodeError(anyhow::format_err!("Duplicate key")));
+            }
+        }
+        Ok(res)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::encoding::tests::test_roundtrip_expected;
+
+    #[test_log::test]
+    fn test_lists() {
+        // The length of the list is encoded before the elements. It is encoded as a
+        // variable length integer, but for lists with a length less than 253, it's
+        // encoded as a single byte.
+        test_roundtrip_expected(&vec![1u8, 2, 3], &[3u8, 1, 2, 3]);
+        test_roundtrip_expected(&vec![1u16, 2, 3], &[3u8, 1, 2, 3]);
+        test_roundtrip_expected(&vec![1u32, 2, 3], &[3u8, 1, 2, 3]);
+        test_roundtrip_expected(&vec![1u64, 2, 3], &[3u8, 1, 2, 3]);
+
+        // Empty list should be encoded as a single byte 0.
+        test_roundtrip_expected::<Vec<u8>>(&vec![], &[0u8]);
+        test_roundtrip_expected::<Vec<u16>>(&vec![], &[0u8]);
+        test_roundtrip_expected::<Vec<u32>>(&vec![], &[0u8]);
+        test_roundtrip_expected::<Vec<u64>>(&vec![], &[0u8]);
+
+        // A length prefix greater than the number of elements should return an error.
+        let buf = [4u8, 1, 2, 3];
+        assert!(
+            Vec::<u8>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default()).is_err()
+        );
+        assert!(
+            Vec::<u16>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default()).is_err()
+        );
+        assert!(
+            VecDeque::<u8>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default())
+                .is_err()
+        );
+        assert!(
+            VecDeque::<u16>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default())
+                .is_err()
+        );
+
+        // A length prefix less than the number of elements should skip elements beyond
+        // the encoded length.
+        let buf = [2u8, 1, 2, 3];
+        assert_eq!(
+            Vec::<u8>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default()).unwrap(),
+            vec![1u8, 2]
+        );
+        assert_eq!(
+            Vec::<u16>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default()).unwrap(),
+            vec![1u16, 2]
+        );
+        assert_eq!(
+            VecDeque::<u8>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default())
+                .unwrap(),
+            vec![1u8, 2]
+        );
+        assert_eq!(
+            VecDeque::<u16>::consensus_decode(&mut buf.as_slice(), &ModuleRegistry::default())
+                .unwrap(),
+            vec![1u16, 2]
+        );
```

Comment:
Note for reviewer: I'm not sure whether this is the behavior we want or not, but these test assertions are revealing behavior that was already there.